From 115326dc4b970eb8f5b424cc88b720a2220f5df2 Mon Sep 17 00:00:00 2001
From: Daniel Badea <daniel.badea@windriver.com>
Date: Fri, 19 Jan 2018 18:32:28 +0000
Subject: [PATCH 15/21] skip checking show_multiple_locations

Allow adding location for rbd images but only if there is no
previous location set. This enables nova snapshot support for
boot-from-image rbd backed instances.
---
 glance/api/v2/images.py | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/glance/api/v2/images.py b/glance/api/v2/images.py
index 029d3a9..5c2ffd8 100644
--- a/glance/api/v2/images.py
+++ b/glance/api/v2/images.py
@@ -422,9 +422,16 @@ class ImagesController(object):
 
     def _do_add_locations(self, image, path_pos, value):
         if CONF.show_multiple_locations == False:
-            msg = _("It's not allowed to add locations if locations are "
-                    "invisible.")
-            raise webob.exc.HTTPForbidden(explanation=msg)
+            # Allow adding location for rbd images but only if there
+            # is no previous location set. This enables nova snapshot support
+            # for boot-from-image rbd backed instances.
+            if value.get('url', '').startswith('rbd://') \
+               and not image.locations:
+                pass
+            else:
+                msg = _("It's not allowed to add locations if locations are "
+                        "invisible.")
+                raise webob.exc.HTTPForbidden(explanation=msg)
 
         if image.status not in ('active', 'queued'):
             msg = _("It's not allowed to add locations if image status is "
-- 
2.7.4

